### ISSUE TITLE
Disabling Grid for TextItem

### ISSUE THREAD

**Issue by Kanonck** [pyqtgraph/pyqtgraph#3072]
Hi,
I am writing on a data-plotting application where I am displaying the graph values in a TextItem Object. I like to display the Grid for my whole Canvas but I don't want the gridlines to overlay on my TextItem, since it makes the values inconvenient to read.

![image](https://github.com/pyqtgraph/pyqtgraph/assets/58548373/b1c44311-824a-4ef6-9fb3-f9af4b010426)

I have already tried:

- setting the fill of the textitem
- setting its z-value very high to be the topmost widget
- using a backgrounditem instead, where I can set opacity

But nothing did what I wanted.
Am I missing something basic? I would imagine there should be a setting to disable gridoverlay for widgets, but I couldn't find anything in the documentation.

Thanks in advace for any help!

**Comment by j9ac9k** [comment]
This issue comes up a lot, and would likely benefit from a documentation change to highlight the workaround, or even better, a redo of GridItem so it works differently.

Basically, you need to change the z-value of the AxisItem, not the GridItem.

Longer answer, GridItem, while a separate GraphicsItem,  for performance reasons was made to be an extension of AxisItem... downside is that you get confusing bits like this.

**Comment by Kanonck** [comment]
Ok, but how do I change that? The Documentation of the AxisItem does not entail a method to change the zValue. Is it related to the Ticks?

**Comment by j9ac9k** [comment]
On mobile so can’t type out a more bullet proof explanation but setZValue
is inherited from Qt, it’s not an AxisItem specific method. The number you
put there determines the order the item is drawn. You want to make AxisItem
have a Z value that is less than the LegendItem.

**Comment by Kanonck** [comment]
Mhm ok, I can call setZValue() on my axis and validate the value by calling myaxis.zValue(). So, now I have set my TextItem.zValue to 10 and my axis zValue to 0.
Still, I can see the grid overlaying on top of my TextItem.
I know that we are at a point where more code is useful, but just to be sure: axis.zValue(0) and TextItem.zValue(10) should not lead to the picture posted earlier, right?

**Comment by j9ac9k** [comment]
oh I would have thought that would do it. Your text box doesn't have a transparent background or something like that does it?

**Comment by Kanonck** [comment]
No, and I found an easy way to demonstrate my problem: the pyqtgraph.examples for TextItem. In this example, one can enable the Grid and then still see the Grid through the TextItem. Even when I change the sourcecode the have the plots zValue(0) and the TextItems zValue(1000) I can see it.
The textitem has no settings for transparency to my knowledge and it is alreayd set to fill its background with the colour.

See the example code with my changes here:

```
"""
This example shows how to insert text into a scene using TextItem. This class 
is for displaying text that is anchored to a particular location in the data
coordinate system, but which is always displayed unscaled. 

For text that scales with the data, use QTextItem. 
For text that can be placed in a layout, use LabelItem.
"""

import numpy as np

import pyqtgraph as pg
from pyqtgraph.Qt import QtCore

x = np.linspace(-20, 20, 1000)
y = np.sin(x) / x
plot = pg.plot()   ## create an empty plot widget
plot.setYRange(-1, 2)
plot.setWindowTitle('pyqtgraph example: text')
plot.setZValue(0)
curve = plot.plot(x,y)  ## add a single curve

## Create text object, use HTML tags to specify color/size
text = pg.TextItem(html='<div style="text-align: center"><span style="color: #FFF;">This is the</span><br><span style="color: #FF0; font-size: 16pt;">PEAK</span></div>', anchor=(-0.3,0.5), angle=45, border='w', fill=(0, 0, 255, 100))
text.setZValue(1000)
plot.addItem(text)
text.setPos(0, y.max())

## Draw an arrowhead next to the text box
arrow = pg.ArrowItem(pos=(0, y.max()), angle=-45)
plot.addItem(arrow)


## Set up an animated arrow and text that track the curve
curvePoint = pg.CurvePoint(curve)
plot.addItem(curvePoint)
text2 = pg.TextItem("test", anchor=(0.5, -1.0))
text2.setParentItem(curvePoint)
arrow2 = pg.ArrowItem(angle=90)
arrow2.setParentItem(curvePoint)

## update position every 10ms
index = 0
def update():
    global curvePoint, index
    index = (index + 1) % len(x)
    curvePoint.setPos(float(index)/(len(x)-1))
    text2.setText('[%0.1f, %0.1f]' % (x[index], y[index]))
    
timer = QtCore.QTimer()
timer.timeout.connect(update)
timer.start(10)

if __name__ == '__main__':
    pg.exec()

```

**Comment by j9ac9k** [comment]
Thanks for posting the example code, when I have some time on my laptop, I'll take a closer look.

**Comment by Silas-K** [comment]
The following solution works for me in the provided example:

- It is interesting to note that setting the `zValue` to `-1` is crucial (`0` does not work).
- The background in the example was indeed slightly transparent, so changing the alpha value from `100` to `255` makes it fully opaque.

```diff
"""
This example shows how to insert text into a scene using TextItem. This class 
is for displaying text that is anchored to a particular location in the data
coordinate system, but which is always displayed unscaled. 

For text that scales with the data, use QTextItem. 
For text that can be placed in a layout, use LabelItem.
"""

import numpy as np

import pyqtgraph as pg
from pyqtgraph.Qt import QtCore

x = np.linspace(-20, 20, 1000)
y = np.sin(x) / x
plot = pg.plot()   ## create an empty plot widget
plot.setYRange(-1, 2)
plot.setWindowTitle('pyqtgraph example: text')
plot.setZValue(0)
curve = plot.plot(x,y)  ## add a single curve

+# Option 1:
+plot.getAxis('left').setZValue(-1)
+plot.getAxis('bottom').setZValue(-1)

+# Option 2:
+# plot.getPlotItem().getAxis('left').setZValue(-1)
+# plot.getPlotItem().getAxis('bottom').setZValue(-1)

## Create text object, use HTML tags to specify color/size
+text = pg.TextItem(html='<div style="text-align: center"><span style="color: #FFF;">This is the</span><br><span style="color: #FF0; font-size: 16pt;">PEAK</span></div>', anchor=(-0.3,0.5), angle=45, border='w', fill=(0, 0, 255, 255))
-text = pg.TextItem(html='<div style="text-align: center"><span style="color: #FFF;">This is the</span><br><span style="color: #FF0; font-size: 16pt;">PEAK</span></div>', anchor=(-0.3,0.5), angle=45, border='w', fill=(0, 0, 255, 100))
text.setZValue(1000)
plot.addItem(text)
text.setPos(0, y.max())

## Draw an arrowhead next to the text box
arrow = pg.ArrowItem(pos=(0, y.max()), angle=-45)
plot.addItem(arrow)


## Set up an animated arrow and text that track the curve
curvePoint = pg.CurvePoint(curve)
plot.addItem(curvePoint)
text2 = pg.TextItem("test", anchor=(0.5, -1.0))
text2.setParentItem(curvePoint)
arrow2 = pg.ArrowItem(angle=90)
arrow2.setParentItem(curvePoint)

## update position every 10ms
index = 0
def update():
    global curvePoint, index
    index = (index + 1) % len(x)
    curvePoint.setPos(float(index)/(len(x)-1))
    text2.setText('[%0.1f, %0.1f]' % (x[index], y[index]))
    
timer = QtCore.QTimer()
timer.timeout.connect(update)
timer.start(10)

if __name__ == '__main__':
    pg.exec()
```
### Comparison

| Before                                                                                   | After                                                                                   |
|------------------------------------------------------------------------------------------|-----------------------------------------------------------------------------------------|
| ![Before](https://github.com/pyqtgraph/pyqtgraph/assets/79999377/9c62ec28-f487-4f67-a380-dbc15cebc8e9) | ![After](https://github.com/pyqtgraph/pyqtgraph/assets/79999377/6386c724-160b-4d9b-a4c0-1fafbb718a5b)  |

**Comment by Kanonck** [comment]
Great, thanks! I can confirm that this works for me too.

**Comment by Kanonck** [comment]
I just found something that might be relevant for future readers. The setAxisItems() method from the PlotItem does set the zValue to 0.5
I realized this because the grid came back overlaying on my textbox after changing something im my plot (which changed the axis). So I always need to reset the zValue again to -1 one after a call to setAxisItems().

Furthermore, the code documentation is misleading, since i states that in order to draw over the axes it is enough to place whatever widget at z>=1 which no enough as we just found in this thread. With axis at z=0.5 and my other widgets at z>=1 it would still draw my axes on top of my widgets.
(see lines 332 and 333 here: https://pyqtgraph.readthedocs.io/en/latest/_modules/pyqtgraph/graphicsItems/PlotItem/PlotItem.html#PlotItem.setAxisItems)

**Comment by j9ac9k** [comment]
Loving the user-to-user support here, as someone that's trying to wrap up a trans content move w/ kids, this kind of communication takes a huge load off for me.

@Kanonck If you're feeling motivated, a PR to update the GridItem docstring with this information would be _greatly_ appreciated.

https://github.com/pyqtgraph/pyqtgraph/blob/master/pyqtgraph/graphicsItems/GridItem.py#L11-L15

Something like adding some sentences describing the issue with overlaying, and the workarounds would be more than sufficient.  Don't worry about formatting.
